### PR TITLE
TST: More old issues

### LINKED ISSUE #19354: Shortcut functions in transform are not grouped
```ipython
$ ipython3
Python 3.6.3 (default, Oct  3 2017, 21:45:48) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.2.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'A': [1, 1, 2], 'B': [1, 2, 3]})

In [3]: df.groupby('A').transform('rank')
Out[3]: 
   B
0  1
1  1
2  2

In [4]: df.groupby('A').transform(lambda x: x.rank())
Out[4]: 
     B
0  1.0
1  2.0
2  1.0

```
#### Problem description
It seems like the functions provided through shorthand strings, such as 'rank', do not obey the groupings of data. For example, one would expect Out[3] and Out[4] to be the same in the code above. Instead, it seems like using `.transform('rank')` doesn't actually obey the grouping and just ranks independently. I've reproduced this with larger dataframes as well.

### LINKED ISSUE #19517: BUG? Possible issue with multi indexing column names of different types
See https://github.com/pandas-dev/pandas/issues/19517#issuecomment-493539602

```python
In [51]: idx1 = pd.MultiIndex.from_product([['a', 'b'], ['A', 'B']])

In [52]: s1 = pd.Series(index=idx1)

In [53]: s1.loc['a']
Out[53]:
A   NaN
B   NaN
dtype: float64

In [54]: idx2 = pd.MultiIndex.from_product([['a', 'b'], ['A', 'B']], names=[1, 0])

In [55]: s2 = pd.Series(index=idx2)

In [56]: s2.loc['a']
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/sandbox/pandas/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2672             try:
-> 2673                 return self._engine.get_loc(key)
   2674             except KeyError:

~/sandbox/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()
    106
--> 107     cpdef get_loc(self, object val):
    108         if is_definitely_invalid_key(val):

~/sandbox/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()
    130         try:
--> 131             return self.mapping.get_item(val)
    132         except (TypeError, ValueError):

~/sandbox/pandas/pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()
   1602
-> 1603     cpdef get_item(self, object val):
   1604         cdef khiter_t k

~/sandbox/pandas/pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()
   1609         else:
-> 1610             raise KeyError(val)
   1611

KeyError: 'a'

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
<ipython-input-56-761f4dad11a7> in <module>
----> 1 s2.loc['a']

~/sandbox/pandas/pandas/core/indexing.py in __getitem__(self, key)
   1502
   1503             maybe_callable = com.apply_if_callable(key, self.obj)
-> 1504             return self._getitem_axis(maybe_callable, axis=axis)
   1505
   1506     def _is_scalar_access(self, key):

~/sandbox/pandas/pandas/core/indexing.py in _getitem_axis(self, key, axis)
   1915         # fall thru to straight lookup
   1916         self._validate_key(key, axis)
-> 1917         return self._get_label(key, axis=axis)
   1918
   1919

~/sandbox/pandas/pandas/core/indexing.py in _get_label(self, label, axis)
    133             # but will fail when the index is not present
    134             # see GH5667
--> 135             return self.obj._xs(label, axis=axis)
    136         elif isinstance(label, tuple) and isinstance(label[axis], slice):
    137             raise IndexingError('no slices here, handle elsewhere')

~/sandbox/pandas/pandas/core/generic.py in xs(self, key, axis, level, drop_level)
   3643         if isinstance(index, MultiIndex):
   3644             loc, new_index = self.index.get_loc_level(key,
-> 3645                                                       drop_level=drop_level)
   3646         else:
   3647             loc = self.index.get_loc(key)

~/sandbox/pandas/pandas/core/indexes/multi.py in get_loc_level(self, key, level, drop_level)
   2587                 return indexer, maybe_droplevels(indexer, ilevels, drop_level)
   2588         else:
-> 2589             indexer = self._get_level_indexer(key, level=level)
   2590             return indexer, maybe_droplevels(indexer, [level], drop_level)
   2591

~/sandbox/pandas/pandas/core/indexes/multi.py in _get_level_indexer(self, key, level, indexer)
   2668         else:
   2669
-> 2670             code = level_index.get_loc(key)
   2671
   2672             if level > 0 or self.lexsort_depth == 0:

~/sandbox/pandas/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2673                 return self._engine.get_loc(key)
   2674             except KeyError:
-> 2675                 return self._engine.get_loc(self._maybe_cast_indexer(key))
   2676         indexer = self.get_indexer([key], method=method, tolerance=tolerance)
   2677         if indexer.ndim > 1 or indexer.size > 1:

~/sandbox/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()
    105         arr[loc] = value
    106
--> 107     cpdef get_loc(self, object val):
    108         if is_definitely_invalid_key(val):
    109             raise TypeError("'{val}' is an invalid key".format(val=val))

~/sandbox/pandas/pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()
    129
    130         try:
--> 131             return self.mapping.get_item(val)
    132         except (TypeError, ValueError):
    133             raise KeyError(val)

~/sandbox/pandas/pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()
   1601                                        sizeof(uint32_t))  # flags
   1602
-> 1603     cpdef get_item(self, object val):
   1604         cdef khiter_t k
   1605

~/sandbox/pandas/pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.get_item()
   1608             return self.table.vals[k]
   1609         else:
-> 1610             raise KeyError(val)
   1611
   1612     cpdef set_item(self, object key, Py_ssize_t val):

KeyError: 'a'

```
---

#### Code Sample, a copy-pastable example if possible

```python
df = pd.read_csv('https://raw.githubusercontent.com/AntoineGautier/Data/master/new.txt')
df.rename({'1': 1, '3': 3, '0': 0}, axis=1, inplace=True)
df.set_index(list(df.columns[:3]), inplace=True)
df.sort_index(inplace=True)
for idx in df.index:
    tmp = df.loc[idx, :]

```
#### Problem description

The code leads to the exception: `KeyError: (1949L, '37,5 cl ', 'Ch\xc3\xa2teau LAFITE ROTHSCHILD ')`.
No exception if all column names are strings.

Remark: This case with mixed types columns names typically occurs when reading from external file with `header=None` and then merging.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.13.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.22.0
pytest: 3.0.7
pip: 9.0.1
setuptools: 38.4.0
Cython: 0.25.2
numpy: 1.14.0
scipy: 0.19.0
pyarrow: None
xarray: None
IPython: 5.5.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.2.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.6.0
html5lib: 0.999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.5.0

</details>

### LINKED ISSUE #20414: API: Inconsistent behavior with setting slices of Series indexed by MultiIndex
#### Code Sample, a copy-pastable example if possible

```python
In [1]: import pandas as pd

In [2]: pd.__version__
Out[2]: '0.22.0'

In [3]: s1 = pd.Series(range(8),
   ...:                index=pd.MultiIndex.from_product([list('ab'),
   ...:                                                  list('xy'),
   ...:                                                  [1,2]],
   ...:                                                  names=['ab','xy','num'])
   ...: )
   ...:

In [4]: s1
Out[4]:
ab  xy  num
a   x   1      0
        2      1
    y   1      2
        2      3
b   x   1      4
        2      5
    y   1      6
        2      7
dtype: int64

In [5]:

In [5]: s2 = pd.Series([100*(i+1) for i in range(4)],
   ...:                index=pd.MultiIndex.from_product([list('ab'),
   ...:                                                  list('xy')],
   ...:                                                  names=['ab','xy']))
   ...:

In [6]: s2
Out[6]:
ab  xy
a   x     100
    y     200
b   x     300
    y     400
dtype: int64

In [7]: s1.loc[pd.IndexSlice[:,:,1]] = -1  # This works as expected

In [8]: s1
Out[8]:
ab  xy  num
a   x   1     -1
        2      1
    y   1     -1
        2      3
b   x   1     -1
        2      5
    y   1     -1
        2      7
dtype: int64

In [9]: s3 = s1.loc[pd.IndexSlice[:,:,1]] + s2 # This works as expected

In [10]: s3
Out[10]:
ab  xy
a   x      99
    y     199
b   x     299
    y     399
dtype: int64

In [11]: s1.loc[pd.IndexSlice[:,:,1]] = s2 # This works differently in v0.22.0 and v0.23 (dev)

In [12]: s1
Out[12]:
ab  xy  num
a   x   1      NaN
        2      1.0
    y   1      NaN
        2      3.0
b   x   1      NaN
        2      5.0
    y   1      NaN
        2      7.0
dtype: float64

In [13]: s1.loc[pd.IndexSlice['a',:,:]] = -2 # This works as expected

In [14]: s1
Out[14]:
ab  xy  num
a   x   1     -2.0
        2     -2.0
    y   1     -2.0
        2     -2.0
b   x   1      NaN
        2      5.0
    y   1      NaN
        2      7.0
dtype: float64

In [15]: s4 = pd.Series([1000*i for i in range(1,5)], index=pd.MultiIndex.from_pr
    ...: oduct([list('xy'),[1,2]], names=['xy','num']))
    ...:

In [16]: s4
Out[16]:
xy  num
x   1      1000
    2      2000
y   1      3000
    2      4000
dtype: int64

In [17]: s5 = s1.loc[pd.IndexSlice['a',:,:]] + s4 # This fails
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-17-cc0933ef6ebd> in <module>()
----> 1 s5 = s1.loc[pd.IndexSlice['a',:,:]] + s4 # This fails

C:\Anaconda3\lib\site-packages\pandas\core\ops.py in wrapper(left, right, name, na_op)
    716             return NotImplemented
    717
--> 718         left, right = _align_method_SERIES(left, right)
    719
    720         converted = _Op.get_op(left, right, name, na_op)

C:\Anaconda3\lib\site-packages\pandas\core\ops.py in _align_method_SERIES(left, right, align_asobject)
    645                 right = right.astype(object)
    646
--> 647             left, right = left.align(right, copy=False)
    648
    649     return left, right

C:\Anaconda3\lib\site-packages\pandas\core\series.py in align(self, other, join, axis, level, copy, fill_value, method, limit, fill_axis, broadcast_axis)
   2605                                          fill_value=fill_value, method=method,
   2606                                          limit=limit, fill_axis=fill_axis,

-> 2607                                          broadcast_axis=broadcast_axis)
   2608
   2609     def rename(self, index=None, **kwargs):

C:\Anaconda3\lib\site-packages\pandas\core\generic.py in align(self, other, join, axis, level, copy, fill_value, method, limit, fill_axis, broadcast_axis)
   5728                                       copy=copy, fill_value=fill_value,
   5729                                       method=method, limit=limit,
-> 5730                                       fill_axis=fill_axis)
   5731         else:  # pragma: no cover
   5732             raise TypeError('unsupported type: %s' % type(other))

C:\Anaconda3\lib\site-packages\pandas\core\generic.py in _align_series(self, other, join, axis, level, copy, fill_value, method, limit, fill_axis)
   5797                 join_index, lidx, ridx = self.index.join(other.index, how=join,
   5798                                                          level=level,
-> 5799                                                          return_indexers=True)
   5800
   5801             left = self._reindex_indexer(join_index, lidx, copy)

C:\Anaconda3\lib\site-packages\pandas\core\indexes\base.py in join(self, other, how, level, return_indexers, sort)
   3101             else:
   3102                 return self._join_multi(other, how=how,
-> 3103                                         return_indexers=return_indexers)
   3104
   3105         # join on the level

C:\Anaconda3\lib\site-packages\pandas\core\indexes\base.py in _join_multi(self, other, how, return_indexers)
   3199                              "overlapping names")
   3200         if len(overlap) > 1:
-> 3201             raise NotImplementedError("merging with more than one level "
   3202                                       "overlap on a multi-index is not "
   3203                                       "implemented")

NotImplementedError: merging with more than one level overlap on a multi-index is not implemented

In [18]: s1.loc[pd.IndexSlice['a',:,:]]  = s4 # This puts in NaN rather than s4 values

In [19]: s1
Out[19]:
ab  xy  num
a   x   1      NaN
        2      NaN
    y   1      NaN
        2      NaN
b   x   1      NaN
        2      5.0
    y   1      NaN
        2      7.0
dtype: float64
```
#### Problem description

This is a bit related to #10440 .  In the above code, if we slice where we fix the value of the third level, then we can change the slice to a constant.  We can also add that slice to a Series that has an Index that matches the first 2 levels.  

In v0.22.0 of pandas, the result of the lines 
```
s1.loc[pd.IndexSlice[:,:,1]] = s2
s1
```
is (as shown above)
```
ab  xy  num
a   x   1      NaN
        2      1.0
    y   1      NaN
        2      3.0
b   x   1      NaN
        2      5.0
    y   1      NaN
        2      7.0
dtype: float64
```
But in the development version 0.23 of pandas, the "correct" result is given:
```
ab  xy  num
a   x   1      100
        2        1
    y   1      200
        2        3
b   x   1      300
        2        5
    y   1      400
        2        7
dtype: int64
```

So I then would expect that the last 2 examples, using `s4` , would work in v0.23 development, because the only difference is that I am fixing the value of the first level in the slice, as opposed to the last level of the slice.  But in both of those cases, I get this error (independent of the pandas version):
```
NotImplementedError: merging with more than one level overlap on a multi-index is not implemented
```

So there is a bit of an inconsistency in that a slice that fixes the last level allows the addition and assignment operations to work (and it is better with v0.23 development version than in v0.22 because the `NaN` values go away), but a slice that fixes the first level does not allow the operations to work.

I'm not sure if this is a bug, or by design, or if the documentation needs to be clarified as to which type of slicing will allow the "setting" operation to work as expected.  There is a line in the docs (http://pandas.pydata.org/pandas-docs/stable/advanced.html#using-slicers) that says "You can use a right-hand-side of an alignable object as well."  At least to me, it's not clear what objects are considered "alignable".

In any case, the expected behavior should be clear in the documentation, and, IMHO, if you fix the value of the first index or the last index, the behavior should be consistent.  

#### Expected Output
```python
In [1]: import pandas as pd

In [2]: pd.__version__
Out[2]: '0.22.0'

In [3]: s1 = pd.Series(range(8),
   ...:                index=pd.MultiIndex.from_product([list('ab'),
   ...:                                                  list('xy'),
   ...:                                                  [1,2]],
   ...:                                                  names=['ab','xy','num'])
   ...: )
   ...:

In [4]: s1
Out[4]:
ab  xy  num
a   x   1      0
        2      1
    y   1      2
        2      3
b   x   1      4
        2      5
    y   1      6
        2      7
dtype: int64

In [5]:

In [5]: s2 = pd.Series([100*(i+1) for i in range(4)],
   ...:                index=pd.MultiIndex.from_product([list('ab'),
   ...:                                                  list('xy')],
   ...:                                                  names=['ab','xy']))
   ...:

In [6]: s2
Out[6]:
ab  xy
a   x     100
    y     200
b   x     300
    y     400
dtype: int64

In [7]: s1.loc[pd.IndexSlice[:,:,1]] = -1  # This works as expected

In [8]: s1
Out[8]:
ab  xy  num
a   x   1     -1
        2      1
    y   1     -1
        2      3
b   x   1     -1
        2      5
    y   1     -1
        2      7
dtype: int64

In [9]: s3 = s1.loc[pd.IndexSlice[:,:,1]] + s2 # This works as expected

In [10]: s3
Out[10]:
ab  xy
a   x      99
    y     199
b   x     299
    y     399
dtype: int64

In [11]: s1.loc[pd.IndexSlice[:,:,1]] = s2 # This works differently in v0.22.0 and v0.23 (dev)

In [12]: s1
Out[12]:
ab  xy  num
a   x   1      NaN
        2      1.0
    y   1      NaN
        2      3.0
b   x   1      NaN
        2      5.0
    y   1      NaN
        2      7.0
dtype: float64

In [13]: s1.loc[pd.IndexSlice['a',:,:]] = -2 # This works as expected

In [14]: s1
Out[14]:
ab  xy  num
a   x   1     -2.0
        2     -2.0
    y   1     -2.0
        2     -2.0
b   x   1      NaN
        2      5.0
    y   1      NaN
        2      7.0
dtype: float64

In [15]: s4 = pd.Series([1000*i for i in range(1,5)], index=pd.MultiIndex.from_pr
    ...: oduct([list('xy'),[1,2]], names=['xy','num']))
    ...:

In [16]: s4
Out[16]:
xy  num
x   1      1000
    2      2000
y   1      3000
    2      4000
dtype: int64

In [17]: s5 = s1.loc[pd.IndexSlice['a',:,:]] + s4 # This should not fail

In[18]: s5
Out[18]:
xy num
x  1   998
   2    1998
y  1    2998
    2    3998

In[19]: s1.loc[pd.IndexSlice['a',:,:]]  = s4 # This should not set NaN

In[20]: s1
ab  xy  num
a   x   1    1000
        2     2000
    y   1    3000
        2     4000
b   x   1      NaN
        2      5.0
    y   1      NaN
        2      7.0
dtype: float64
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.22.0
pytest: 3.3.2
pip: 9.0.1
setuptools: 38.4.0
Cython: 0.27.3
numpy: 1.14.0
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.6.6
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.2
openpyxl: 2.4.10
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.1
pymysql: 0.7.11.None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### LINKED ISSUE #20420: Inconsistent groupby-apply output shape and random values returned.
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame({"A": [1, 2, 3, 4, 5], "B": [6, 7, 8, 9, 0],
                   "C": [1, 1, 1, 2, 2]}, index=range(5))     
a = df.groupby("C").apply(lambda x: x.A)                      
b = df.groupby("C").apply(lambda x: x.A.sort_index())

In [9]: print(df)
   A  B  C
0  1  6  1
1  2  7  1
2  3  8  1
3  4  9  2
4  5  0  2

In [248]: print(a)
C
1  0    1
   1    2
   2    3
2  3    4
   4    5
Name: A, dtype: int64

In [249]: print(b)
A  0  1   2
C
1  1  2   3
2  4  5  33
```
#### Problem description

First the output of .groupby().apply() seems inconsistent, sometimes it returns the "correct" shape as in case A while in case B the output is transposed.

Second, the 33 returned values in case B is not what I would expect it to be. That number changes if I call the function multiple times.

This does not only happens when sort_index() is called but it was the simplest example I could consistently reproduce.

#### Expected Output

```python
C
1  0    1
   1    2
   2    3
2  3    4
   4    5
Name: A, dtype: int64
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-112-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.22.0
pytest: 3.2.1
pip: 9.0.1
setuptools: 36.5.0.post20170921
Cython: 0.26.1
numpy: 1.13.3
scipy: 1.0.0
pyarrow: 0.8.0
xarray: None
IPython: 6.1.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: None
numexpr: None
feather: None
matplotlib: 2.1.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.5.0
</details>

### LINKED ISSUE #20520: data frame group by with extra categorical column aggregates datetime64 NaT column to float NaN
#### Code Sample, a copy-pastable example if possible

```python
# Minimal example:
df = pd.DataFrame({'group': ['first', 'first', 'second', 'third', 'third'], 'time': 5*[np.datetime64('NaT')], 'categories': pd.Series(["a","b","c","a","b"], dtype="category")})
df.groupby('group').first()
# results in
>>> 	       categories	time
>>> group		
>>> first       a                NaN
>>> second	c	         NaN
>>> third	a	         NaN

# while
df = pd.DataFrame({'group': ['first', 'first', 'second', 'third', 'third'], 'time': 5*[np.datetime64('NaT')]})
df.groupby('group').first()
# results in
>>> 	      time
>>> group		
>>> first     NaT
>>> second    NaT
>>> third     NaT
```
#### Problem description

When aggregating a NaT valued datetime64 data frame column in the presence of a second categorical column the resulting aggregated column is a NaN valued float. This does not happen in the absence of the categorical column.

#### Expected Output

   	       categories	time
  group		
  first       a                NaT
  second	c	         NaT
  third	a	         NaT

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.9.final.0
python-bits: 64
OS: Linux
OS-release: 4.9.0-0.bpo.4-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.22.0
pytest: None
pip: 9.0.1
setuptools: 28.8.0
Cython: None
numpy: 1.14.0
scipy: 1.0.0
pyarrow: 0.8.0
xarray: None
IPython: 5.5.0
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2018.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: 1.2.2
pymysql: None
psycopg2: 2.7.3.2 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### LINKED ISSUE #20734: Creating DataFrame throws: data type "bytes512" not understood
#### Code Sample, a copy-pastable example if possible

```python
index = pd.Series(name='id', dtype='S24')
df = pd.DataFrame(index=index)
df['a'] = pd.Series(name='a', index=index, dtype=np.uint32)
df['b'] = pd.Series(name='b', index=index, dtype='S64')
df['c'] = pd.Series(name='c', index=index, dtype='S64')
df['d'] = pd.Series(name='d', index=index, dtype=np.uint8)
```

#### Problem description

The code, above, which is attempting to create an empty `pandas.DataFrame` with an index and four typed columns yields the following error:

```
[... snip ...]\appdata\local\programs\python\python36\lib\site-packages\pandas\core\internals.py in _vstack(to_stack, dtype)
   4912 
   4913     # work around NumPy 1.6 bug
-> 4914     if dtype == _NS_DTYPE or dtype == _TD_DTYPE:
   4915         new_values = np.vstack([x.view('i8') for x in to_stack])
   4916         return new_values.view(dtype)

TypeError: data type "bytes512" not understood
```

Why?

Changing the order of the columns works just fine:

```python
index = pd.Series(name='id', dtype='S24')
df = pd.DataFrame(index=index)
df['a'] = pd.Series(name='a', index=index, dtype=np.uint32)
df['d'] = pd.Series(name='d', index=index, dtype=np.uint8)
df['b'] = pd.Series(name='b', index=index, dtype='S64')
df['c'] = pd.Series(name='c', index=index, dtype='S64')
```

In fact, it seems that any `Series` added after the two `S64` series throws an error: I tried with both `np.float` and `np.bool`.

#### Expected Output

I would expect that it isn't important which order the Series are added or, if it actually *is* important, perhaps a better error message.

I tried with an older version of Python 3.6, NumPy and Pandas and then updated, thinking this was just a bug. The latest version I tested was CPython 3.6.5, NumPy 1.14.2, Pandas 0.22.0.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 26 Stepping 5, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.22.0
pytest: None
pip: 9.0.3
setuptools: 39.0.1
Cython: None
numpy: 1.14.2
scipy: 1.0.1
pyarrow: None
xarray: None
IPython: 6.3.1
sphinx: None
patsy: None
dateutil: 2.7.2
pytz: 2018.4
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #20951: Addressing multiindex raises TypeError if indices that are rightmost are not present
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

A = [100, 100, 200, 200, 300, 300]
B = [10, 10, 20, 21, 31,33]
C = np.random.randint(0, 99, 6)

test_df = pd.DataFrame({'A': A, 'B': B, 'C': C})
test_df = test_df.set_index(['A', 'B'])

print(test_df)
try:
    print(test_df.loc[(100,10)])
except:
    pass

try:
    print(test_df.loc[(0,1)])
except KeyError:
    print('test_df.loc[(0,1)] raises a KeyError')

try:
    print(test_df.loc[(100,1)])
except KeyError as e:
    print(e)
    print('test_df.loc[(100,1)] raises a KeyError')
except TypeError as e:
    print(e)
    print('test+df.loc[(100,1)]) raises a TypeError')

```
#### Problem description
If the value is not present in the index, I believe that a KeyError should be raised consistently, so you can write code like.
```python
try:
   df.loc[tuple]
except KeyError:
   # do something if the value isn't present
```
#### Expected Output
If df.loc[tuple] does not have a match in the multiindex, a KeyError should be raised.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.22.0
pytest: None
pip: 9.0.3
setuptools: 39.0.1
Cython: None
numpy: 1.14.2
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.2
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #21651: BUG: Pandas groupby datetime and column then apply generates ValueError
#### Code Sample, a copy-pastable example if possible
```
    | SYM_ROOT | TIME_M                     | BEST_BID | BEST_ASK | increment | genjud_incre | 
    |----------|----------------------------|----------|----------|-----------|--------------| 
    | A        | 2017-01-03 09:30:00.004712 | 45.91    | 46.12    | 0         | 4680         | 
    | AA       | 2017-01-03 09:30:00.004014 | 28.55    | 28.57    | 0         | 4680         | 
```
```python
# Your code here
df=pd.read_csv('stack.csv')
df['TIME_M']=pd.to_datetime(df['TIME_M'],format='%Y%m%d %H:%M:%S.%f')
df.groupby(['SYM_ROOT',df['TIME_M'].dt.date]).apply(group_increment_to_end)

def group_increment_to_end(x):
    return x.iloc[0:1]
```
#### Problem description

I am trying to group my dataframe and then apply a function to each row of the dataframe. SYM_ROOT is a category variable, while TIME_M is a datetime variable.

However, I keep getting the following error:

`ValueError: Key 2017-01-03 00:00:00 not in level Index([2017-01-03], dtype='object', name=u'TIME_M')`

I am referring to [this](https://stackoverflow.com/questions/51057324/pandas-groupby-datetime-and-column-then-apply-generates-valueerror/51058220#51058220) stackoverflow post. 

I think the reason is that pandas somehow doesn't recognize the date object in the column correctly when using the group by statement. It falsely think compares `2017-01-03 00:00:00` with `TIME_M` and generates the error. After separating out the date as a single column, the problem fixes itself. 

But I think it will be beneficial if pandas can recognize the date object correctly in the columns ...

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.14.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-693.21.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_CA.UTF-8
LOCALE: None.None

pandas: 0.22.0
pytest: 3.3.2
pip: 9.0.1
setuptools: 38.4.0
Cython: 0.27.3
numpy: 1.14.0
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 5.4.1
sphinx: 1.6.6
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.2
openpyxl: 2.4.10
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.1
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #19354
- [x] closes #19517
- [x] closes #20414
- [x] closes #20420
- [x] closes #20520
- [x] closes #20734
- [x] closes #20951
- [x] closes #21651
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

### CODE REVIEW

File: pandas/tests/indexing/multiindex/test_loc.py [hunk 1]
```diff
@@ -866,3 +866,27 @@ def test_loc_get_scalar_casting_to_float():
...
+
+    df = DataFrame(
+        {
+            "A": [100, 100, 200, 200, 300, 300],
+            "B": [10, 10, 20, 21, 31, 33],
+            "C": range(6),
+        }
+    )
+    df = df.set_index(["A", "B"])
+    with pytest.raises(KeyError, match="1"):
```
Comment (simonjayhawkins) [inline pandas/tests/indexing/multiindex/test_loc.py:891]:
```suggestion
    with pytest.raises(KeyError, match="^1$"):
```

**Review by simonjayhawkins — APPROVED** [review]
Thanks @mroeschke lgtm. one suggestion.

**Review by jreback — APPROVED** [review]
sure, ex @simonjayhawkins comment

**Comment by jreback** [comment]
thanks @mroeschke

**Pull request merged**